Modernize post: Creating Plugins for the Cockpit User Interface (#316)
cockpit-project/cockpit-project.github.io: _posts/2014-11-13-cockpit-plugin-tutorial.md

---
title: Creating Plugins for the Cockpit User Interface
date: '2020-05-04'
category: tutorial
tags: cockpit linux
slug: creating-plugins-for-the-cockpit-user-interface
---

> *Note: This post has been updated on May 2020 to adjust to current Cockpit UI and development best practices.*

Reviewer: ```suggestion
_**Note**: This post was updated in May 2020 to adjust to reflect Cockpit UI and development best practices._
```

> [Cockpit is a user interface for servers](https://cockpit-project.org). You can easily add your own pages to that user interface, i. e. components that appear in the menu:

Reviewer: ```suggestion
[Cockpit is a web-based graphical interface for servers](https://cockpit-project.org). You can easily add your own custom pages to the navigation.
```

![Navigation menu](/images/pages-menu-top.png)

> Lets make ourselves another one. For this tutorial you need to [install your distribution's cockpit packages](../running.html) or [build it from git](https://github.com/cockpit-project/cockpit/blob/master/HACKING.md).

Reviewer: ```suggestion
For this tutorial you need to [install your distribution's cockpit packages](../running.html) or [build it from git](https://github.com/cockpit-project/cockpit/blob/master/HACKING.md).
```

I don't understand what "Lets make ourselves another one." is supposed to mean. (It's supposed to be "Let's make another one (_of something_) ourselves." if used.)

Perhaps: "Let's add our own page now." Or just drop the sentence entirely, as it's already implied by "You can easily add your own custom pages to the navigation." directly above.

Author: Agreed, that's the point of the entire blog post, after all. :-)

> So break out your terminal, lets make a package called *pinger* that checks whether your server has network connectivity to the Internet by pinging another host. Nothing too fancy. We'll just be spawning a process on the server to do the work. This package is already [included in the Cockpit sources as an example](https://github.com/cockpit-project/cockpit/tree/master/examples/pinger). You can look it over and modify it.

Reviewer: ```suggestion
We'll make a package called *pinger* that checks whether your server has network connectivity to the Internet by pinging another host. It's simple and not too fancy. The package will spawn a process on the server to do all the work.

This example package is already [included in the Cockpit sources](https://github.com/cockpit-project/cockpit/tree/master/examples/pinger). You can look it over and modify it.

To start, let's get ready for development by launching a terminal on your local computer. 
```

> To download the example to your current directory:

Reviewer: ```suggestion
First, create a project directory and download the example:
```

```sh
mkdir pinger; cd pinger
curl -O https://raw.githubusercontent.com/cockpit-project/cockpit/master/examples/pinger/manifest.json
curl -O https://raw.githubusercontent.com/cockpit-project/cockpit/master/examples/pinger/ping.html
curl -O https://raw.githubusercontent.com/cockpit-project/cockpit/master/examples/pinger/pinger.js
```

> Cockpit pages, and more specifically their HTML and Javascript files, live in [package directories](https://cockpit-project.org/guide/latest/packages.html). In the package directory there's also a [manifest.json](https://cockpit-project.org/guide/latest/packages.html#package-manifest) file which tells Cockpit about the package. The `pinger` directory above is such a package. Its `manifest.json` file looks like this:

Reviewer: ```suggestion
Cockpit pages — more specifically their HTML and Javascript files — live in [package directories](https://cockpit-project.org/guide/latest/packages.html). In the package directory there's also a [manifest.json](https://cockpit-project.org/guide/latest/packages.html#package-manifest) file which tells Cockpit about the package. The `pinger` directory above is such a package. Its `manifest.json` file looks like this:
```

            "label": "Pinger",
            "path": "ping.html"
        }
    }
}
```

The manifest above has a `"tools"` subsection, which means that it will appear in the lower section of the menu (see the [manifest documentation](https://cockpit-project.org/guide/latest/packages.html#package-manifest) for details). Each tool is listed in the menu by Cockpit. The `"path"` is the name of the HTML file that implements the tool, and the `"label"` is the text to show in the menu.

> You'll notice that we haven't told Cockpit about how to find the package yet. To do so you either place or symlink the package into one of two places:

Reviewer: ```suggestion
You'll notice that we haven't told Cockpit about how to find the package yet. To do so you either move or symlink the package into one of two places:
```

Reviewer: We probably want something more specific than "place" here. I'm not sure if "move" is right... but perhaps it is?

Author: "copy", I think. I'll add you suggestion and squash in the s/move/copy/.

>  * `~/.local/share/cockpit` in your home directory, for user specific packages, and ones that you're working on. You can edit these on the fly and just refresh your browser to see changes.

Reviewer: ```suggestion
 * `~/.local/share/cockpit` is in your home directory. It's used for user specific packages and ones that you're developing. You can edit these on the fly and refresh your browser to see changes.
```

Author: I'd drop the first "is", it's confusing IMHO.

>  * `/usr/share/cockpit` for installed packages available to all users. These should not be changed while Cockpit is running.

Reviewer: ```suggestion
 * `/usr/share/cockpit` is the location for installed packages available to all users of a system. Changing files in this path requires administrator ("root") privileges. These should not be changed while Cockpit is running.
```

> Since we're going to be messing around with this package, lets symlink it into the former location.

Reviewer: ```suggestion
Since we're going to be actively editing this package, let's symlink it into the first location, in your home directory.
```

```sh
mkdir -p ~/.local/share/cockpit
ln -snf $PWD ~/.local/share/cockpit/pinger
```

> You can list which Cockpit packages are installed using the following command, and you should see `pinger` listed among them:

Reviewer: ```suggestion
To list the Cockpit packages which are installed, use the following command:
```

...

> ```

Reviewer: ```suggestion
```

You should see `pinger` listed among all the active modules, like the above example.
```

> Now log into Cockpit on this machine, or if you already are, refresh the page. Make sure to log into Cockpit with your current user name, since you installed the package in your home directory. You should now see a new item:

Reviewer: ```suggestion
Log into Cockpit on this machine with your current user name, as the package is installed in your home directory. (If you're already logged in to Cockpit with your user account, you can simply reload your browser.)

You should now see a new item:
```

    <title>Pinger</title>
    <meta charset="utf-8">
    <link href="../base1/cockpit.css" type="text/css" rel="stylesheet">
    <script src="../base1/cockpit.js"></script>
</head>
<body>
    <div class="container-fluid">
        <table class="form-table-ct">
            <tr>
                <td><label class="control-label" for="address">Address</label></td>
                <td><input class="form-control" id="address" value="8.8.8.8"></td>
            </tr>
            <tr>
                <td><button class="pf-c-button pf-m-primary" id="ping">Ping</button></td>
                <td><span id="result"></span></td>
            </tr>
        </table>
        <pre id="output"></pre>
    </div>

    <script src="pinger.js"></script>
</body>
</html>

```

> The page includes the cockpit.css file to make sure the look of our tool matches that of Cockpit.

Reviewer: ```suggestion
The `cockpit.css` file is included so our tool matches the style of the rest of Cockpit.
```

> Then we include `cockpit.js`: the basic API for interacting with the system, as well as Cockpit itself. You can find [detailed documentation here](https://cockpit-project.org/guide/latest/api-cockpit.html).

Reviewer: ```suggestion
Here we include `cockpit.js`: the basic API for interacting with the system, as well as Cockpit itself. You can find [detailed documentation in the Cockpit guide](https://cockpit-project.org/guide/latest/api-base1.html).
```

(Changed the link to point to the contents above the page, as the `api-cockpit` page is mostly empty and the `api-base1` page is mainly documentation on `cockpit.js`.)

> The HTML is pretty basic, defining a little form with a field to type an address, a button to click to start the pinging, and an area to present output and results.

Reviewer: ```suggestion
The HTML is pretty basic. It defines a little form with a field to type an address, a button to click to start pinging, and an area to present output and results.
```

```js
const address = document.getElementById("address");
const output = document.getElementById("output");
const result = document.getElementById("result");
const button = document.getElementById("ping");

function ping_run() {
    cockpit.spawn(["ping", "-c", "4", address.value])
        .stream(ping_output)
        .then(ping_success)
        .catch(ping_fail);

    result.innerHTML = "";
    output.innerHTML = "";
}

function ping_success() {
    result.style.color = "green";
    result.innerHTML = "success";
}

function ping_fail() {
    result.style.color = "red";
    result.innerHTML = "fail";
}

function ping_output(data) {
    output.append(document.createTextNode(data));
}

// Connect the button to starting the "ping" process
button.addEventListener("click", ping_run);

// Send a 'init' message.  This tells integration tests that we are ready to go
cockpit.transport.wait(function() { });
```

First we get a bunch of variables pointing to the HTML elements we want to interact with. Next we attach a handler to the *Ping* button so that the `ping_run()` function is called when it is clicked.

```js
function ping_run() {
}
...
button.addEventListener("click", ping_run);
```

> In the `ping_run()` function is where the magic happens. [cockpit.spawn](https://cockpit-project.org/guide/latest/cockpit-spawn.html) lets you spawn processes on the server and interact with them via stdin and stdout. Here we spawn the `ping` command with some arguments:

Reviewer: ```suggestion
The `ping_run()` function is where the magic happens. [cockpit.spawn](https://cockpit-project.org/guide/latest/cockpit-spawn.html) lets you spawn processes on the server and interact with them via `stdin` and `stdout`.

Here we spawn the `ping` command with some arguments:
```

```js
cockpit.spawn(["ping", "-c", "4", address.val()])
```

In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/cockpit-spawn.html#cockpit-spawn-then).

> JavaScript has a standard interface called a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). You add handlers by calling the `.then()` or `.catch()` methods and registering callbacks (Note: historically these have been called `.done()` and `.fail()`, but these should not be used any more in new code.) The `cockpit.spawn` specific `.stream()` handler registers a callback to be invoked whenever the process produces output.

Reviewer: Should we have something around these paragraphs that have an important / warning alert that says it's important that you cannot depend on the browser session to be constant, due to _reasons_ (and list the reasons)?

It would help prevent issues with people coding pages that somehow expect 2-hour-long continuous browser sessions.

Author: That's a good idea, I added a warning.

```js
   <some promise>
   .stream(ping_output)
   .then(ping_success)
   .catch(ping_fail);
}
```

The `ping_success()` and `ping_fail()` and `ping_output()` update the display as you would expect.

> This should be enough for your first useful (but admittedly ugly) Cockpit page!  Please see the [Contributing](https://cockpit-project.org/external/wiki/Contributing.html) page for more links where to go from here.

Reviewer: ```suggestion
This should be enough to get you started with your first useful (but admittedly basic) Cockpit page!  Please see the [Contributing](https://cockpit-project.org/external/wiki/Contributing.html) page for more documentation on where to grow from here.
```

Is changing go to grow here too "cute"? As they're going to be growing their experience and their own custom pages, I think it's OK. But would like feedback to be sure.

Also, I changed "ugly" to "basic". It's probably better to say it this way, and should be more PC ("politically correct", not "personal computer") and friendly.

>     <title>Pinger</title>

Reviewer: GitHub won't let me make a suggestion for a collapsed block. Bummer.

Anyway, the HTML is missing an doctype and `<html>` tag at the beginning.

Please add the following to the top of the HTML block:

```
<!doctype html>
<html lang="en">
```

(Also please check to make sure this is in the actual file that's downloaded too.)

Author: Indeed, I sent https://github.com/cockpit-project/cockpit/pull/14029 to fix that. I'll already include it in my follow-up.

Here we spawn the `ping` command with some arguments:

```js
cockpit.spawn(["ping", "-c", "4", address.val()])
```

In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/cockpit-spawn.html#cockpit-spawn-then).

JavaScript has a standard interface called a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). You add handlers by calling the `.then()` or `.catch()` methods and registering callbacks (Note: historically these have been called `.done()` and `.fail()`, but these should not be used any more in new code.) The `cockpit.spawn` specific `.stream()` handler registers a callback to be invoked whenever the process produces output.

```js
   <some promise>
   .stream(ping_output)
   .then(ping_success)
   .catch(ping_fail);
}
```

The `ping_success()` and `ping_fail()` and `ping_output()` update the display as you would expect.

> **Warning**: Don't start long-running and uninterruptible processes in this naïve way. Browser tabs get closed, internet connections severed, laptops get suspended, etc. Take care to use a mechanism like [systemd-run](https://www.freedesktop.org/software/systemd/man/systemd-run.html) or similar for use cases like installation processes that should not be interrupted by a closing Cockpit session.

Reviewer: ```suggestion
_**Warning**: Don't start long-running, uninterruptible tasks in this naïve way. Browser tabs are easily closed, Internet connections get severed, laptops suspend, and so on. Take care to use a mechanism like [systemd-run](https://www.freedesktop.org/software/systemd/man/systemd-run.html) (or similar) for processes that should not be interrupted, such as installation procedures._
```

Reviewer: We could also use an admonition style here too (in addition to the note at the top of the page).

Don't let either lack of call-out style block this PR though. We should open another issue to handle it and retroactively adjust these (and possibly other blog posts) "note:", "info:", "warning:", etc. notices.
